### PR TITLE
[AIR/Train] Add `Trainer.restore` API for train experiment-level fault tolerance

### PR DESCRIPTION
**Pull request by justinvyu**
## Why are these changes needed?

This PR introduces a `Trainer.restore` API that resumes a Train experiment that crashed/got interrupted. Previously, `resume_from_checkpoint` only allowed starting a _new_ experiment, which writes to a completely new log directory and a new set of results.

### Context
The restoration API is inconsistent between Train and Tune, since `Tuner.restore`. It’s not clear how to restore an interrupted Train experiment, and the existing `resume_from_checkpoint` API has issues as reported by users (see here [[github](https://github.com/ray-project/ray/issues/28829), [discourse](https://discuss.ray.io/t/train-when-resuming-training-a-new-trial-directory-is-created-even-when-resuming-from-checkpoint/7672), [github](https://github.com/ray-project/ray/issues/25441)]).

### Added Documentation

- [FAQ on restoring a Train experiment](https://ray--31920.org.readthedocs.build/en/31920/train/faq.html#how-do-i-restore-a-ray-train-experiment)
- [API reference](https://ray--31920.org.readthedocs.build/en/31920/train/api/api.html#trainer-restore)
- [Docstrings](https://ray--31920.org.readthedocs.build/en/31920/train/api/doc/ray.train.trainer.BaseTrainer.restore.html#ray.train.trainer.BaseTrainer.restore)

### TODO before merging
- [x] Merge Jun's PR, overwrite the param space, and do the shutdown/init in the unit test to make sure that it doesn't hang/crash.
- [x] Add to docs
- [x] Fix restore for with user-defined Trainer subclasses that are defined in the same scope as the `trainer.fit` call.

#### Dealing with `ActorHandle`
TL;DR: Object references don't immediately throw an exception when unpickled from a new ray cluster, but actor handles do.

If an `ActorHandle` gets captured in the scope of a training function/config and gets pickled along with the trainer, then it will be impossible to load the pickled trainer from a new ray cluster. You will run into an error that looks like:

```
Traceback (most recent call last):
  File "train_example_nfs.py", line 85, in <module>
    trainer = DataParallelTrainer.restore(
  File "/Users/justin/Developer/justinvyu-dev/python/ray/train/data_parallel_trainer.py", line 315, in restore
    return super(DataParallelTrainer, cls).restore(
  File "/Users/justin/Developer/justinvyu-dev/python/ray/train/base_trainer.py", line 281, in restore
    original_trainer = pickle.load(fp)
  File "/Users/justin/Developer/justinvyu-dev/python/ray/_private/serialization.py", line 89, in _actor_handle_deserializer
    return ray.actor.ActorHandle._deserialization_helper(serialized_obj, outer_id)
  File "/Users/justin/Developer/justinvyu-dev/python/ray/actor.py", line 1280, in _deserialization_helper
    return worker.core_worker.deserialize_and_register_actor_handle(
  File "python/ray/_raylet.pyx", line 2310, in ray._raylet.CoreWorker.deserialize_and_register_actor_handle
  File "python/ray/_raylet.pyx", line 2279, in ray._raylet.CoreWorker.make_actor_handle
  File "/Users/justin/Developer/justinvyu-dev/python/ray/_private/function_manager.py", line 522, in load_actor_class
    actor_class = self._load_actor_class_from_gcs(
  File "/Users/justin/Developer/justinvyu-dev/python/ray/_private/function_manager.py", line 617, in _load_actor_class_from_gcs
    class_name = ensure_str(class_name)
  File "/Users/justin/Developer/justinvyu-dev/python/ray/_private/utils.py", line 293, in ensure_str
    assert isinstance(s, bytes)
```
We can't really do anything about this, and this de-serialization in a new ray cluster behavior needs to be fixed by Core.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/base_trainer.py [hunk 1]
```diff
@@ -351,25 +410,41 @@ def fit(self) -> Result:
...
 
-        tuner = Tuner(
-            trainable=trainable, param_space=param_space, run_config=self.run_config
-        )
+        if self._restore_path:
+            tuner = Tuner.restore(
+                self._restore_path,
+                overwrite_trainable=trainable,
+                resume_unfinished=True,
+                resume_errored=True,
```
Comment (gjoliver) [inline python/ray/train/base_trainer.py:423]: why set this to True? machine goes away will cause all the error trials to be retried is a bit weird?

File: python/ray/train/base_trainer.py [hunk 2]
```diff
@@ -400,18 +475,32 @@ def _generate_trainable_cls(self) -> Type["Trainable"]:
...
+            checkpoint = session.get_checkpoint()
+            loaded_preprocessor = False
+            if checkpoint:
+                trainer.resume_from_checkpoint = checkpoint
+                # Always use the checkpointed preprocessor when retrying on errors
+                # (configured by FailureConfig(max_failures)).
+                # On a restored the full experiment via Trainer.restore, only load the
+                # checkpointed preprocessor if the preprocessor is not re-specified.
+                should_load_from_checkpoint = (
+                    not trainer._restore_path or not trainer.preprocessor
```
Comment (gjoliver) [inline python/ray/train/base_trainer.py:494]: name this ``should_load_preprocessors_from_checkpoint``?
also the logic seems weird, we will do that if _restore_path is None, regardless of trainer.preprocessor state?

File: python/ray/train/huggingface/huggingface_trainer.py [hunk 3]
```diff
@@ -283,6 +281,58 @@ def __init__(
...
+        if TRAINER_INIT_FN_KEY in trainer_init_config:
+            raise ValueError(
+                f"'{TRAINER_INIT_FN_KEY}' is a reserved key in `trainer_init_config`."
+            )
+        if trainer_init_per_worker:
+            trainer_init_config[TRAINER_INIT_FN_KEY] = trainer_init_per_worker
+        return trainer_init_config
+
+    @classmethod
+    def restore(
```
Comment (gjoliver) [inline python/ray/train/huggingface/huggingface_trainer.py:303]: it kinda bothers me a lot now that we have to create a custom restore() for every single Trainer we have.
nobody would understand how to contribute a new Trainer if we do this ...

File: python/ray/train/BUILD [hunk 4]
```diff
@@ -219,321 +296,537 @@ py_test(
...
-    deps = [":train_lib", ":conftest"]
+    tags = [
+        "exclusive",
+        "ray_air",
+        "team:ml",
+    ],
+    deps = [
+        ":conftest",
+        ":train_lib",
+    ],
```
[the remark this answers is not in the document]
Comment (justinvyu) [reply python/ray/train/BUILD:688]: Yes, can revert this. I have some format-on-save thing that did this.

File: python/ray/train/base_trainer.py (hunk 1 above)
Comment (justinvyu) [reply python/ray/train/base_trainer.py:423]: There is only ever 1 trial in the Trainer case. Do you think that this is not desired behavior on restore?

I'm thinking about the case where: Trainer errors due to some transient error, causing the script to fail. Job retries, and we want to resume the errored run.

File: python/ray/train/base_trainer.py (hunk 2 above)
Comment (justinvyu) [reply python/ray/train/base_trainer.py:494]: Yep you're right, should just be `should_load_preprocessor_from_checkpoint = not trainer.preprocessor`.

File: python/ray/train/huggingface/huggingface_trainer.py (hunk 3 above)
Comment (justinvyu) [reply python/ray/train/huggingface/huggingface_trainer.py:303]: What about:
Allow user to re-specify everything in `__init__` except for the run config. We can inspect the arguments of `cls.__init__` to see what's allowed. Then just overwrite everything that's passed in. Now, the subclasses don't need to re-implement restore.

One downside is that validation will be harder, and the signature of restore will be something like `def restore(cls, **kwargs):` which is very opaque.

File: python/ray/train/base_trainer.py [hunk 5]
```diff
@@ -278,7 +365,7 @@ def setup(self) -> None:
         """
         pass
 
-    def preprocess_datasets(self) -> None:
+    def preprocess_datasets(self, should_fit_preprocessor: bool = True) -> None:
```
Comment (krfricke) [inline python/ray/train/base_trainer.py:368]: Can we shorten this name? I think this is expressive enough
```suggestion
    def preprocess_datasets(self, fit_preprocessor: bool = True) -> None:
```

File: python/ray/train/base_trainer.py (hunk 1 above)
Comment (krfricke) [reply python/ray/train/base_trainer.py:451]: I agree with Justin, I think `resume_errored=True` is desired here

File: python/ray/train/huggingface/huggingface_trainer.py [hunk 6]
```diff
@@ -283,6 +284,27 @@ def __init__(
...
+            trainer_init_config[TRAINER_INIT_FN_KEY] = self.trainer_init_per_worker
+        return trainer_init_config
+
+    @train_loop_config.setter
+    def train_loop_config(self, config):
+        # This disables setting the `train_loop_config` attribute.
+        # The `train_loop_config` for this Trainer depends only on
+        # the `trainer_init_per_worker` and `trainer_init_config` args,
+        # and will be created with the getter above.
+        pass
```
Comment (krfricke) [inline python/ray/train/huggingface/huggingface_trainer.py:306]: Do we call this somewhere? Or is this just if the user calls it? Should we raise a warning or an error if this is called?

File: python/ray/train/base_trainer.py [hunk 7]
```diff
@@ -164,8 +175,68 @@ def __init__(
...
+        trainer.preprocessor = preprocessor
+
+        if scaling_config:
+            trainer.scaling_config = scaling_config
+
+        for field_name in cls._optional_fields_for_restore:
+            original_attr = getattr(trainer, field_name, None)
+            respecified = field_name in kwargs
+            if contains_object_refs(original_attr) and not respecified:
+                raise ValueError()
```
Comment (krfricke) [inline python/ray/train/base_trainer.py:234]: We probably want to fill this out

File: python/ray/util/check_object_refs.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+
+def contains_object_refs(base_obj: Union[Dict, Sequence, Callable]) -> bool:
+    if base_obj is None:
+        return False
+
+    object_store_types = (ObjectRef, ActorHandle)
+
+    if isinstance(base_obj, dict):
+        return any(isinstance(v, object_store_types) for v in base_obj.values())
```
Comment (krfricke) [inline python/ray/util/check_object_refs.py:43]: I think we need to call this recursively to catch nested dicts

**Review by krfricke — COMMENTED** [review]
Generally I think the approach is good

File: python/ray/train/huggingface/huggingface_trainer.py (hunk 6 above)
Comment (justinvyu) [reply python/ray/train/huggingface/huggingface_trainer.py:306]: This is called by us in the super class constructor. This was a hack to get restore to work for the Huggingface/Mosaic trainers, without needing to implement `HuggingFaceTrainer.restore`.

Relates to Jun's comment here: https://github.com/ray-project/ray/pull/31920#discussion_r1087122528

File: python/ray/train/base_trainer.py (hunk 7 above)
Comment (justinvyu) [reply python/ray/train/base_trainer.py:234]: Yep, going to do a pass to fill in all the errors!

File: python/ray/train/_internal/dataset_spec.py [hunk 9]
```diff
@@ -141,8 +143,15 @@ def preprocess_datasets(
                     continue
                 if conf.fit:
                     ds_to_fit = datasets[k]
-            if ds_to_fit:
+            if ds_to_fit and prep.fit_status() in (
+                Preprocessor.FitStatus.NOT_FITTED,
+                Preprocessor.FitStatus.PARTIALLY_FITTED,
+            ):
```
Comment (justinvyu) [inline python/ray/train/_internal/dataset_spec.py:149]: Is this a controversial change? A loaded preprocessor that's been fitted already shouldn't fit again.

File: python/ray/train/base_trainer.py [hunk 10]
```diff
@@ -351,25 +413,41 @@ def fit(self) -> Result:
...
         assert len(result_grid) == 1
-        try:
-            result = result_grid[0]
-            if result.error:
-                raise result.error
-        except TuneError as e:
-            raise TrainingFailedError from e
+        result = result_grid[0]
+        if result.error:
+            raise TrainingFailedError from result.error
```
Comment (justinvyu) [inline python/ray/train/base_trainer.py:443]: I changed this because the old try/catch did not most errors raised by the trial, and I was asserting for `TrainingFailedError` in my tests.

Comment (justinvyu) [reply python/ray/train/base_trainer.py]: Reverted this. Will do this in a separate PR.

File: python/ray/train/base_trainer.py [hunk 11]
```diff
@@ -351,25 +413,41 @@ def fit(self) -> Result:
...
+                trainable=trainable, param_space=param_space, run_config=self.run_config
+            )
+            experiment_path = Path(
+                TunerInternal.setup_create_experiment_checkpoint_dir(
+                    trainable, self.run_config
+                )
+            )
+            self._save(experiment_path)
+
         result_grid = tuner.fit()
```
[the remark this answers is not in the document]
Comment (justinvyu) [reply python/ray/train/base_trainer.py:574]: Will do this in another PR.

File: python/ray/train/huggingface/huggingface_trainer.py (hunk 3 above)
Comment (justinvyu) [reply python/ray/train/huggingface/huggingface_trainer.py:306]: Latest iteration: User subclass doesn't need to set any attributes themselves. They just need to define what attributes from their constructor can get re-specified. Then, they just pass these through as kwargs to the `BaseTrainer.restore`.

```python
class MyTrainerSubclass(BaseTrainer):
    def restore(cls, path, my_arg, ...):
        return super().restore(path, my_arg=my_arg, ...)
        # No more setting things afterwards.
```

File: python/ray/train/base_trainer.py [hunk 12]
```diff
@@ -300,7 +489,10 @@ def preprocess_datasets(self) -> None:
 
         if self.preprocessor:
             train_dataset = self.datasets.get(TRAIN_DATASET_KEY, None)
-            if train_dataset:
+            if train_dataset and self.preprocessor.fit_status in (
+                Preprocessor.FitStatus.NOT_FITTED,
+                Preprocessor.FitStatus.PARTIALLY_FITTED,
+            ):
```
Comment (gjoliver) [inline python/ray/train/base_trainer.py:495]: for my education, can you share a bit why we have this logic both here and in dataset_spec.py?

File: python/ray/train/huggingface/huggingface_trainer.py (hunk 3 above)
Comment (gjoliver) [reply python/ray/train/huggingface/huggingface_trainer.py:306]: this is nice!

File: python/ray/train/tests/test_trainer_restore.py [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+    if trainer_cls == HuggingFaceTrainer:
+        from ray.train.tests.test_huggingface_trainer import (
+            train_function as hf_init,
+            train_df,
+        )
+
+        trainer_init_fn = hf_init
+        trainer_init_config = {"epochs": 5, "save_strategy": "epoch"}
+        datasets = {"train": ray.data.from_pandas(train_df)}
+    # TODO(ml-team): Add MosaicTrainer test after Mosaic checkpointing is supported
```
Comment (gjoliver) [inline python/ray/train/tests/test_trainer_restore.py:177]: what does this comment mean?

**Review by gjoliver — APPROVED** [review]

File: python/ray/train/tests/test_trainer_restore.py (hunk 13 above)
Comment (justinvyu) [reply python/ray/train/tests/test_trainer_restore.py:177]: MosaicTrainer doesn't perform checkpointing at all right now and needs to be added. So, restore/`resume_from_checkpoint` will not do anything.

File: python/ray/train/base_trainer.py (hunk 12 above)
Comment (justinvyu) [reply python/ray/train/base_trainer.py:495]: The `DataParallelTrainer` goes through a separate path for preprocessing datasets. It uses a data parallel ingest spec. Other trainers like xgboost will let the underlying framework handle the data ingest (ex: RayDMatrix params can be passed in to the underlying `xgboost_ray` object.)

File: python/ray/train/tests/test_trainer_restore.py (hunk 13 above)
Comment (gjoliver) [reply python/ray/train/tests/test_trainer_restore.py:177]: oh I see. I didn't notice we are testing multiple trainers in a single test.
maybe just separate them, and create 1 restore test for each trainer?

Comment (justinvyu) [reply python/ray/train/tests/test_trainer_restore.py:177]: I am grouping tests with similar trainers (ex: xgboost/lightgbm, huggingface/mosaic) where only a few things need to be changed, and they will still show up as 2 different tests in the CI. Do you think that's okay?

Comment (gjoliver) [reply python/ray/train/tests/test_trainer_restore.py:177]: ah sure, sure 👍

File: python/ray/train/base_trainer.py [hunk 14]
```diff
@@ -164,8 +176,169 @@ def __init__(
...
+                modified to be different from the original spec.
+            **kwargs: Other optionally re-specified arguments, passed in by subclasses.
+
+        Raises:
+            ValueError: If all datasets were not re-supplied on restore.
+
+        Returns:
+            BaseTrainer: A restored instance of the class that is calling this method.
+        """
+        assert cls.can_restore(path), (
```
Comment (Yard1) [inline python/ray/train/base_trainer.py:271]: Can we avoid using assertions and instead raise Value/Runtime errors?

File: python/ray/train/base_trainer.py [hunk 15]
```diff
@@ -164,8 +176,169 @@ def __init__(
...
+        original_datasets = original_trainer.datasets or {}
+        if original_datasets and not datasets:
+            raise ValueError(
+                "The following datasets need to be provided again on restore: "
+                f"{list(original_datasets.keys())}\n"
+                f"Use {cls.__name__}.restore(..., datasets=datasets) "
+                "with the datasets that were provided to the original trainer."
+            )
+        datasets = datasets or {}
+        assert set(original_datasets.keys()) == set(datasets.keys()), (
```
Comment (Yard1) [inline python/ray/train/base_trainer.py:300]:
```suggestion
        assert set(original_datasets) == set(datasets), (
```

File: python/ray/train/base_trainer.py [hunk 16]
```diff
@@ -164,8 +176,169 @@ def __init__(
...
+
+        Returns:
+            bool: Whether or not this path exists and contains the pickled Trainer
+        """
+        path = str(path)
+        if is_non_local_path_uri(path):
+            dir_contents = list_at_uri(path)
+        else:
+            dir_contents = [] if not os.path.exists(path) else os.listdir(path)
+        return bool(dir_contents) and _TRAINER_PKL in dir_contents
```
Comment (Yard1) [inline python/ray/train/base_trainer.py:340]:
```suggestion
        return dir_contents and _TRAINER_PKL in dir_contents
```

BTW. this is the same as in the other PR, right? Should we make it a separate AIR-level function that is simply wrapped around by Trainer & Tuner classmethods?

File: python/ray/train/base_trainer.py [hunk 17]
```diff
@@ -351,15 +543,34 @@ def fit(self) -> Result:
...
+            tuner = Tuner(
+                trainable=trainable, param_space=param_space, run_config=self.run_config
+            )
+
+        experiment_path = Path(
+            TunerInternal.setup_create_experiment_checkpoint_dir(
+                trainable, self.run_config
+            )
         )
+        self._save(experiment_path)
```
Comment (Yard1) [inline python/ray/train/base_trainer.py:572]: can we add a comment explaining what this does?

File: python/ray/train/base_trainer.py (hunk 14 above)
Comment (justinvyu) [reply python/ray/train/base_trainer.py:271]: Got it, asserts should only be used for internal sanity checks. Not external validation.

File: python/ray/train/base_trainer.py (hunk 16 above)
Comment (justinvyu) [reply python/ray/train/base_trainer.py:340]: Yes, can do.

File: python/ray/train/base_trainer.py [hunk 18]
```diff
@@ -164,8 +176,169 @@ def __init__(
...
+        path: str,
+        datasets: Optional[Dict[str, GenDataset]] = None,
+        preprocessor: Optional["Preprocessor"] = None,
+        scaling_config: Optional[ScalingConfig] = None,
+        **kwargs,
+    ) -> "BaseTrainer":
+        """Restores a Train experiment from a previously interrupted/failed run.
+
+        Restore should be used for experiment-level fault tolerance in the event
+        that the head node crashes (e.g. OOM or some other runtime error) or the
```
Comment (angelinalg) [inline python/ray/train/base_trainer.py:196]:
```suggestion
        that the head node crashes (e.g., OOM or some other runtime error) or the
```

File: python/ray/train/base_trainer.py [hunk 19]
```diff
@@ -164,8 +176,169 @@ def __init__(
...
+        entire cluster goes down (e.g. network error affecting all nodes).
```
Comment (angelinalg) [inline python/ray/train/base_trainer.py:197]:
```suggestion
        entire cluster goes down (e.g., network error affecting all nodes).
```

File: python/ray/train/base_trainer.py [hunk 20]
```diff
@@ -164,8 +176,169 @@ def __init__(
...
+        trainer._restore_path = path
+        return trainer
+
+    @classmethod
+    def can_restore(cls: Type["BaseTrainer"], path: Union[str, Path]) -> bool:
+        """Checks whether a given directory contains a restorable Train experiment.
+
+        Args:
+            path: The path to the experiment directory of the Train experiment.
+                This can be either a local directory (e.g. ~/ray_results/exp_name)
```
Comment (angelinalg) [inline python/ray/train/base_trainer.py:329]:
```suggestion
                This can be either a local directory (e.g., ~/ray_results/exp_name)
```

File: python/ray/train/base_trainer.py [hunk 21]
```diff
@@ -164,8 +176,169 @@ def __init__(
...
+                or a remote URI (e.g. s3://bucket/exp_name).
```
Comment (angelinalg) [inline python/ray/train/base_trainer.py:330]:
```suggestion
                or a remote URI (e.g., s3://bucket/exp_name).
```

File: python/ray/train/data_parallel_trainer.py [hunk 22]
```diff
@@ -283,6 +280,47 @@ def __init__(
...
+        datasets: Optional[Dict[str, GenDataset]] = None,
+        preprocessor: Optional["Preprocessor"] = None,
+        scaling_config: Optional[ScalingConfig] = None,
+    ) -> "DataParallelTrainer":
+        """Restores a DataParallelTrainer from a previously interrupted/failed run.
+
+        Args:
+            train_loop_per_worker: Optionally re-specified train loop function.
+                This should be used to re-specify a function that is not
+                restorable in a new Ray cluster (e.g. it holds onto outdated
```
Comment (angelinalg) [inline python/ray/train/data_parallel_trainer.py:300]:
```suggestion
                restorable in a new Ray cluster (e.g., it holds onto outdated
```

**Review by krfricke — COMMENTED** [review]
Looks good to me - let's wait for Jun's PR

**Review by krfricke — APPROVED** [review]

**Review by krfricke — APPROVED** [review]
Thanks! Please ping me when tests pass
